### ISSUE TITLE
disconnects after login with hardened kernel and xorgxrdp

### ISSUE THREAD

**Issue by joshiggins** [neutrinolabs/xrdp#885]
I've seen a lot of issues lately regarding session terminates just before the desktop appears. There are some patterns: startwm.sh is good, it always occurs with Xorg session, and other session types will work fine like Xvnc.

I don't know exactly how the communication between X server and xrdp (through xorgxrdp module) is implemented, but on hardened kernels that restrict insecure IPC it does not work. This issue arises with SELinux and grsecurity depending on policy configuration.

I'm running [xrdp in Docker](https://github.com/kxes/desktop) and the problem is resolved by adding the IPC_OWNER capability.

The [grsecurity docs](https://en.wikibooks.org/wiki/Grsecurity/Appendix/Grsecurity_and_PaX_Configuration_Options#Disallow_access_to_overly-permissive_IPC_objects) say that access to IPC will be denied for a process

```
1) If the IPC object is world-accessible and the euid doesn't match
   that of the creator or current uid for the IPC object
2) If the IPC object is group-accessible and the egid doesn't
   match that of the creator or current gid for the IPC object
```

So, the Xorg process is started as the user that has been authenticated, and then it is accessed from xrdp process that is running as root. This is denied by the policy.

Confirmed via syslog, message appears at exact moment RDP client disconnects (after login, but before desktop is shown)

```
Sep 21 14:27:56 foo kern.alert kernel: [ 113.764641] grsec: From 10.0.2.2: denied write of overly-permissive IPC object with creator uid 1000 by /usr/local/sbin/xrdp[xrdp:2257] uid/euid:0/0 gid/egid:0/0, parent /usr/local/sbin/xrdp[xrdp:2211] uid/euid:0/0 gid/egid:0
```

I would hesitantly suggest that distros shipping with SELinux or grsecurity enforcing out of the box will probably result in the same IPC denial - fedora, alpine, etc. 

It is easy to reproduce - just set up an Alpine VM, `apk add docker` and then run `docker run --net=host kxes/desktop`. Connect to it (the default login is user/changeme) and the session will disconnect right after login - check the syslog on the host.

There are a few options:

### 1) turn off insecure IPC protection

With grsecurity, this is easy

```
sysctl -w kernel.grsecurity.harden_ipc=0
```

Not sure with SELinux, but setting to permissive mode does the trick.

### 2) request CAP_IPC_OWNER for xrdp daemon

This would ignore IPC permission checking on the xrdp process itself, thus allowing it access to any other process shared memory and whatever, if I understood the docs correctly.

### 3) fix the permissions

Not sure if this would even be possible. I assume there is a good reason why the X server is run as user, which is why there is also no option 4 :wink:

---

Sorry for the long post, this is my first contribution. Thanks for making such a cool thing!

**Comment by jsorg71** [comment]
There is a good reason Xorg is run as user, yes.
So we have xrdp(root) <-> Xorg(user)
I've been considering to drop the xrdp process to the user after authentication.
Would that solve this problem?
I think some packaged xrdp from distros do not startup as root so we can't setuid is that case.

**Comment by joshiggins** [comment]
I can't think of anything that would work in the non-root case. It would not be possible to setuid or request the IPC_OWNER capability unless it was running as root as you said.

Maybe the best solution is up to packagers. It should be possible to make an SELinux policy I suppose.

Alternatively, systemd could give the CAP_IPC_OWNER (using `CapabilityBoundingSet=` in unit file) to xrdp and still make the effective uid non-root when starting the service. In this case, the capability wouldn't be inherited by child processes started by xrdp so nothing else would be exempt.

**Comment by joshiggins** [comment]
However, since we are messing with capabilities, with CAP_SETUID and CAP_SETGID it is possible to make arbitrary manipulations of the effective user and group without running as root..... only work on Linux though.

**Comment by matt335672** [comment]
I believe this is fixed for v0.10.x, as we've moved to POSIX shared memory for IPC rather than using System-V.
